iot-e2e-tests/common/src/test/java/tests/integration/com/microsoft/azure/sdk/iot/helpers/Tools.java

```diff
@@ -139,6 +153,99 @@ public static Device addDeviceWithRetry(RegistryManager registryManager, Device
         return ret;
     }
 
+    public static void addDevicesWithRetry(List<Device> devices, String connectionString) throws IotHubException, IOException, InterruptedException
+    {
+        // IoT hub only allows for bulk adding of devices at up to 100 per request, so take the provided devices iterable
+        // and break it into 100 devices or smaller chunks
+        List<Device> subIterable = new ArrayList<>();
+        List<Device> devicesClone = new ArrayList<>(); //create a clone of the source list so elements can be removed from it instead
+        devicesClone.addAll(devices);
+        while (devicesClone.size() > 0)
+        {
+            Device device = devicesClone.remove(0);
+            subIterable.add(device);
+
+            // wait until sub list has the Iot Hub limit of 100 devices to add, or until there will be no more devices to add
+            if (subIterable.size() > 99 || devicesClone.size() <= 0)
+            {
+                long startTime = System.currentTimeMillis();
+                while (System.currentTimeMillis() - startTime < RETRY_TIMEOUT_ON_NETWORK_FAILURE_MILLISECONDS)
+                {
+                    try
+                    {
+                        addDevices(subIterable, connectionString);
+                        break;
+                    }
+                    catch (UnknownHostException | SocketException | SocketTimeoutException e)
```

Comment:
Well that is something interesting that Java can do

Reply:
I like it, personally